label_studio/io_storages/base_models.py

```diff
@@ -551,15 +554,16 @@ def save_all_annotations(self):
         self.info_set_in_progress()
         self.cached_user = self.project.organization.created_by
 
-        for annotation in Annotation.objects.filter(project=self.project).iterator(
-            chunk_size=settings.STORAGE_EXPORT_CHUNK_SIZE
-        ):
-            annotation.cached_user = self.cached_user
-            self.save_annotation(annotation)
+        max_workers = min(32, (os.cpu_count() or 4) * 4)
 
-            # update progress counters
-            annotation_exported += 1
-            self.info_update_progress(last_sync_count=annotation_exported, total_annotations=total_annotations)
+        with ThreadPoolExecutor(max_workers=max_workers) as executor:
+            futures = []
+            for annotation in Annotation.objects.filter(project=self.project).iterator(chunk_size=settings.STORAGE_EXPORT_CHUNK_SIZE):
```

Comment:
notably, we have to submit all annots here before we update any progress, so I'm planning to just do this in batches (using the same batch size we use to fetch from the db), so we can incrementally update the progress incrementally.
We could also update the progress in the thread, and I may experiment with that if I have time, but it seems a bit simpler to just batch the updates imo